### PR TITLE
task-router: Wire up to execution and scheduler servers

### PR DESCRIPTION
**Pull request by bduffany**
Depends on https://github.com/buildbuddy-io/buildbuddy/pull/511

* Call `router.RankNodes` from scheduler server (replacing the old `SampleNodes`)
* Call `router.MarkTaskComplete` from execution server (to update the routing table after each executed task)
* Provision the enterprise server with a task router
* Add integration test showing that we can create a file and read it back from the same workspace (in separate tasks) even though there are multiple executors that could receive the read task.
  * Sanity checked that it fails if we don't pass `recycle-runner=true`, meaning that the task router is doing its job.
  * Made sure to seed random with the current time, for extra confidence that it's not just passing because the seed happens to assign the task to the same executor twice.
---

**Version bump**: Minor

### CODE REVIEW

File: enterprise/server/remote_execution/execution_server/execution_server.go [hunk 1]
```diff
@@ -771,3 +784,39 @@ func (s *ExecutionServer) PublishOperation(stream repb.Execution_PublishOperatio
...
+	router := s.env.GetTaskRouter()
+	if router == nil {
+		return nil
+	}
+	// If the result was served from cache, nothing was actually executed, so no
+	// need to update the task router.
+	if executeResponse.GetCachedResult() {
+		return nil
+	}
+	// TODO: Maybe pre-fetch the command at the beginning of PublishOperation,
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/execution_server/execution_server.go:798]: you can prolly use readProtoFromCAS here instead

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go [hunk 2]
```diff
@@ -147,6 +149,26 @@ func (en *executionNode) GetExecutorID() string {
...
+	for _, node := range nodes {
+		en, ok := node.(*executionNode)
+		if !ok {
+			return nil, status.InternalError("failed to convert executionNode to interface; this should never happen")
+		}
+		out = append(out, en)
+	}
+	return out, nil
+}
+
```
Comment (tylerwilliams) [inline enterprise/server/scheduling/scheduler_server/scheduler_server.go:171]: Why are these necessary / is there a better way?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/remote_execution/execution_server/execution_server.go (hunk 1 above)
Comment (bduffany) [reply enterprise/server/remote_execution/execution_server/execution_server.go:793]: Done -- had to attach the user prefix to the context at the beginning of `PublishOperation` to get it to work (I guess when going through the BS API like I was doing before, that happens automatically?). Also cleaned up the `readProtoFromCAS` method a bit -- looks like we can directly call `cachetools.ReadProtoFromCAS` now

File: enterprise/server/remote_execution/execution_server/execution_server.go [hunk 3]
```diff
@@ -324,15 +324,7 @@ func (s *ExecutionServer) operationFromExecution(ctx context.Context, te *tables
 }
 
 func (s *ExecutionServer) readProtoFromCAS(ctx context.Context, d *digest.InstanceNameDigest, msg proto.Message) error {
```
Comment (vadimberezniker) [inline enterprise/server/remote_execution/execution_server/execution_server.go:326]: nit: Should we remove this function and call cachetools.ReadProtoFromCAS directly?
There are only two calls that need to be updated.

File: enterprise/server/remote_execution/execution_server/execution_server.go [hunk 4]
```diff
@@ -730,6 +725,19 @@ func (s *ExecutionServer) PublishOperation(stream repb.Execution_PublishOperatio
...
 
+		// Before publishing the COMPLETED operation, update the task router
+		// so that subsequent tasks with similar routing properties can get
+		// routed back to the same executor.
+		if stage == repb.ExecutionStage_COMPLETED {
+			response := operation.ExtractExecuteResponse(op)
+			if response != nil {
+				if err := s.updateRouter(ctx, taskID, response); err != nil {
+					// Errors from updating the router should not be fatal.
+					log.Errorf("Failed to update task router: %s", err)
```
Comment (vadimberezniker) [inline enterprise/server/remote_execution/execution_server/execution_server.go:736]: Can you add the taskID to the error message (unless the returned error already includes it)

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go [hunk 5]
```diff
@@ -1186,12 +1188,21 @@ func (s *SchedulerServer) enqueueTaskReservations(ctx context.Context, enqueueRe
...
+				if opts.alwaysScheduleLocally {
+					nodes = nodeBalancer.connectedExecutors
+				}
+				if len(nodes) == 0 {
 					return status.UnavailableErrorf("No registered executors in pool %q with os %q with arch %q.", pool, os, arch)
 				}
-				nodes = newNodes
+				rankedNodes, err := s.taskRouter.RankNodes(ctx, cmd, remoteInstanceName, toNodeInterfaces(nodes))
+				if err != nil {
+					log.Errorf("Failed to rank nodes: %s", err)
```
Comment (vadimberezniker) [inline enterprise/server/scheduling/scheduler_server/scheduler_server.go:1200]: What do you think about changing the RankNodes interface to not return an error and instead having the task router log a message instead? Doesn't seem like there's anything the calling code can do with the error and it's a bit confusing to both return a valid list of nodes and an error.

**Review by vadimberezniker — APPROVED** [review]
LGTM

File: enterprise/server/remote_execution/execution_server/execution_server.go (hunk 3 above)
Comment (bduffany) [reply enterprise/server/remote_execution/execution_server/execution_server.go]: Done

File: enterprise/server/remote_execution/execution_server/execution_server.go (hunk 4 above)
Comment (bduffany) [reply enterprise/server/remote_execution/execution_server/execution_server.go]: Done

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go (hunk 5 above)
Comment (bduffany) [reply enterprise/server/scheduling/scheduler_server/scheduler_server.go]: SG, done. I was originally thinking that it's nice for tests to be able to check the err, but it seems just as effective to only check the nodes that are returned.

File: enterprise/server/remote_execution/execution_server/execution_server.go [hunk 6]
```diff
@@ -771,3 +775,35 @@ func (s *ExecutionServer) PublishOperation(stream repb.Execution_PublishOperatio
...
+	router := s.env.GetTaskRouter()
+	if router == nil {
+		return nil
+	}
+	// If the result was served from cache, nothing was actually executed, so no
+	// need to update the task router.
+	if executeResponse.GetCachedResult() {
+		return nil
+	}
+	// TODO: Maybe pre-fetch the command at the beginning of PublishOperation,
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/execution_server/execution_server.go:789]: nit: can you remove this comment? (It's kinda 'speculative')

File: enterprise/server/remote_execution/execution_server/execution_server.go (hunk 1 above)
Comment (tylerwilliams) [reply enterprise/server/remote_execution/execution_server/execution_server.go:789]: Yup, makes sense. When you were using the bytestream client before you were going through the external cache impl which does that internally.

File: enterprise/server/remote_execution/execution_server/execution_server.go [hunk 7]
```diff
@@ -771,3 +775,35 @@ func (s *ExecutionServer) PublishOperation(stream repb.Execution_PublishOperatio
...
+	actionInstanceNameDigest := digest.NewInstanceNameDigest(d, instanceName)
+	action := &repb.Action{}
+	if err := cachetools.ReadProtoFromCAS(ctx, s.cache, actionInstanceNameDigest, action); err != nil {
+		return status.WrapError(err, "get action from CAS")
+	}
+	cmdDigest := action.GetCommandDigest()
+	cmdInstanceNameDigest := digest.NewInstanceNameDigest(cmdDigest, instanceName)
+	cmd := &repb.Command{}
+	if err := cachetools.ReadProtoFromCAS(ctx, s.cache, cmdInstanceNameDigest, cmd); err != nil {
+		return status.WrapError(err, "get command from CAS")
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/execution_server/execution_server.go:804]: please remove wrapError calls (and take a look at https://github.com/buildbuddy-io/buildbuddy/pull/534)

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go [hunk 8]
```diff
@@ -147,6 +149,29 @@ func (en *executionNode) GetExecutorID() string {
 	return en.executorID
 }
 
+// TODO: Expand interfaces.ExecutionNode interface to include all executionNode
```
Comment (tylerwilliams) [inline enterprise/server/scheduling/scheduler_server/scheduler_server.go:155]: nit: please assign this TODO to a person or bug. A la:

TODO(bduffs): do the thing OR
TODO(b/1w123123): do the thing

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go (hunk 2 above)
Comment (tylerwilliams) [reply enterprise/server/scheduling/scheduler_server/scheduler_server.go:177]: Thanks for the explanation!

That interface change seems like a great way to clean this up, do you plan to do that in a followup?

**Review by tylerwilliams — APPROVED** [review]
LGTM (with a few nits)

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go (hunk 2 above)
Comment (bduffany) [reply enterprise/server/scheduling/scheduler_server/scheduler_server.go:177]: Yep, I will work on that after merging this one.

File: enterprise/server/remote_execution/execution_server/execution_server.go (hunk 6 above)
Comment (bduffany) [reply enterprise/server/remote_execution/execution_server/execution_server.go]: Done

File: enterprise/server/remote_execution/execution_server/execution_server.go (hunk 7 above)
Comment (bduffany) [reply enterprise/server/remote_execution/execution_server/execution_server.go]: Done
